### PR TITLE
switch contract storage to child trie

### PR DESCRIPTION
**Pull request by cheme**
This PR switch double key storage for contract to child trie storage see  #1864.

This PR stays on the area of the initial issue by trying to be small and unobtrusive, but I added some note about things that I think should be done (I started the PR from the db side first then switch to this version that replies more directly to the initial issue).

There is a major issue/tradeoff in this PR : I currently do not touch child trie implementation so the indirection is stored in a different place than the trie root at a big cost:
- trieid stored at accountid (with account runtime storage macro gen prefix)
- childtrie root stored at trieid (with child trie prefix)
- keyval stored at key val from childtrie root

This is needed to allow switching trie (related to #2035 (key isolation cannot be made from child trie address or trie root)).
Alternate design would be to store the uniqueid (equivalent to trieid) at the same place as actual childtrie root, allowing to switch place of a subtrie without the indirection.  Things will be:
- childtrie root and copy of accontid(or independant uniqueid) at accountid (with child trie prefix)
- keyval stored at key val from childtrie root
 
Edit: As I removed child trie changes from this PR, I am also removing the discussion about the possibility of child trie generalization (making root a child trie and going in a hierarchical trie model).

### CODE REVIEW

**Comment by xlc** [comment]
(Sorry this may not make sense because my knowledge of trie storage is very limited)
Is it possible to implement `StorageDoubleMap` with child trie storage?
The main advantage of `StorageDoubleMap` over `Map<(Key1Type, Key2Type), ValueType>` is support batch deletion of `Key1Type` prefix and I need this property in my parachain runtime module.
Or does the current implementation already supports efficient batch deletion?

**Comment by cheme** [comment]
`StorageDoubleMap` could live independently of child trie, it is basically a way of storing in the trie at a prefixed location (the hash of the first key). Therefore it allows deletion of all the element prefixed with the first key without messing to much with the trie structure (being prefixed in a first key long enough to avoid hash collision).
Child trie store the root hash of a second trie in a parent trie at a place corresponding to `Key1` of the `StorageDoubleMap`, so deletion means removing this root hash (single deletion) from parent trie and dropping the entire child trie (full trie deletion): that is pretty similar (for the parent trie it is a single leaf deletion so it is similar as cutting off a prefixed branch).

For efficient deletion, it is already efficient at the trie level (comparable with double storage implementation code).

Optimization up to the db level is a different question and requires changing pruning code to allow dropping a full child trie, this is also impacting a lot more code (many temporary storage for pending changes). This is not ready for prefix deletion either (`StorageDoubleMap` do not have this optimization either). 

TLDR; this is similar and in a similar optimized state (but child trie got some things to fix first), you can definitely use a `StorageDoubleMap` into a child trie so it makes sense to me to keep the code (even if there may be no use case after the contract storage switch)

**Comment by gavofyork** [comment]
Would like @arkpar's opinion on that. Storage-double-map usage could ultimately lead to the need for ref counting or other non-performant ways of mitigating the fact that the same node may be placed in the trie multiple times (with different prefixes).

File: srml/contract/src/account_db.rs [hunk 1]
```diff
@@ -83,23 +140,24 @@ impl<T: Trait> AccountDb<T> for DirectAccountDb {
...
+					child::kill(&keyspace[..], &k);
 				}
 			}
 		}
 	}
 }
-
 pub struct OverlayAccountDb<'a, T: Trait + 'a> {
 	local: RefCell<ChangeSet<T>>,
+	keyspace_account: Option<RefCell<AccountKeySpaceMapping<<T as system::Trait>::AccountId>>>,
```
Comment (pepyakin) [inline srml/contract/src/account_db.rs]: Why do we need this kind of chaining of `AccountKeySpaceMapping`? Can we simplify it, e.g. by using a single instance of it?

File: srml/contract/src/tests.rs [hunk 2]
```diff
@@ -111,6 +116,15 @@ impl ContractAddressFor<H256, u64> for DummyContractAddressFor {
 	}
 }
 
+pub struct DummyKeySpaceGenerator;
+impl KeySpaceGenerator<u64> for DummyKeySpaceGenerator {
+	fn key_space(account_id: &u64) -> KeySpace {
+		let mut res = Vec::new();
+		res.extend_from_slice(&account_id.to_be_bytes()[..]);
+		res
```
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/tests.rs]: yes, bad code from refactoring, a previous version was using more parameters (on multilayer subtrie I used an additional reference to top keyspace).

File: core/primitives/src/storage.rs [hunk 3]
```diff
@@ -79,6 +79,9 @@ pub mod well_known_keys {
 	/// Changes trie configuration is stored under this key.
 	pub const CHANGES_TRIE_CONFIG: &'static [u8] = b":changes_trie";
 
+	/// Prefix of child storage keys.
+	pub const CONTRACT_SUBTRIE: &'static [u8] = b":contract_subtrie:";
```
[the remark this answers is not in the document]
Comment (cheme) [reply core/primitives/src/storage.rs]: Originaly I was using the child_trie_prefix location, I still want (see pr comment) to fuse the storage for SubTrie infos with the trie_root and trie_root is not managed at a contract level (so decl_storage seems awkward).
Reading your comment about mem_size I think we will need this kind of struct anyway, so the change makes sense (will do the switch)

File: srml/contract/src/lib.rs [hunk 4]
```diff
@@ -92,6 +93,50 @@ pub trait ComputeDispatchFee<Call, Balance> {
...
 }
 
+#[derive(Encode,Decode,Clone,Debug)]
+/// Information for managing a sub trie abstraction.
+/// This is the required info to cache for an account
+pub struct SubTrie {
+	/// unique ID for the subtree encoded as a byte
+	pub key_space: KeySpace,
+	/// the size of stored value in octet
+	pub current_mem_stored: u64,
```
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/lib.rs]: Yes this 'SubTrie' can be the 'AccountDbInfo' runtime version of account storage infos.
Still I need a similar desk for actual 'SubTrie', with trie root, unique db keyspace (not memory usage as if we want different storage strategy depending on mem usage at db level that would be the db impl that would track it). But this depends on where we want to go with child trie.

File: srml/contract/src/lib.rs [hunk 5]
```diff
@@ -342,11 +389,14 @@ decl_storage! {
...
+		/// The subtrie counter associated with an account id
+		pub AccountCounter: u64 = 0;
 	}
 }
 
-/// The storage items associated with an account/key.
-///
+/// The storage items associated with a prefix space in a tree
+/// TODO can we remove it, seems usefull for inner implementation
+/// just not in the right place
```
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/lib.rs]: It is an almost working implementation (from Gav last remark I see that it really need at least a small fix on full_key calculation to have the key2 part made unique from key1).
So yes there is no reason, just a tool that may prove being useful for other runtime.

File: srml/contract/src/lib.rs [hunk 6]
```diff
@@ -92,6 +93,50 @@ pub trait ComputeDispatchFee<Call, Balance> {
...
+		let new_seed = <AccountCounter<T>>::get().wrapping_add(1);
+
+		let mut buf = Vec::new();
+		buf.extend_from_slice(account_id.as_ref());
+		buf.extend_from_slice(&new_seed.to_be_bytes()[..]);
+		// this line is yelling use Hash instead of vec: would require making KeySpace a subtype of
+		// Trait
+		let res = T::Hashing::hash(&buf[..]).as_ref().into();
+	
+		<AccountCounter<T>>::put(new_seed);
```
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/lib.rs]: I like committed only on the end of the method since there is no rollback in case of error (no reason for error at this point). But it is true that worst case scenario we are skipping a value that does not need to be sequential. So I can use that (second TODO after decl_storage).

File: srml/executive/src/lib.rs [hunk 7]
```diff
@@ -309,8 +309,8 @@ mod tests {
 		type Hash = substrate_primitives::H256;
 		type Hashing = BlakeTwo256;
 		type Digest = Digest;
-		type AccountId = u64;
-		type Lookup = IdentityLookup<u64>;
+		type AccountId = u64; // TODO only u64??
```
[the remark this answers is not in the document]
Comment (cheme) [reply srml/executive/src/lib.rs]: removed, sorry some comment from my earlier code reading.

File: srml/contract/src/tests.rs [hunk 8]
```diff
@@ -111,6 +116,15 @@ impl ContractAddressFor<H256, u64> for DummyContractAddressFor {
 	}
 }
 
+pub struct DummyKeySpaceGenerator;
+impl KeySpaceGenerator<u64> for DummyKeySpaceGenerator {
+	fn key_space(account_id: &u64) -> KeySpace {
```
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/tests.rs]: I was looking at the contractid dummy method (a few line above) and thought that it would not need it.
Should we also consider changing `DummyContractAddressFor`?

File: core/state-machine/src/lib.rs [hunk 9]
```diff
@@ -870,6 +873,65 @@ mod tests {
 		assert_eq!(ext.child_storage(b":child_storage:testchild", b"abc"), None);
 	}
 
+	// #[test] TODO this is currently broken it requires key isolation for child storage
+	fn child_storage_keyspace() {
```
[the remark this answers is not in the document]
Comment (cheme) [reply core/state-machine/src/lib.rs]: I need to create the issue first :) but sure.

File: srml/contract/src/account_db.rs [hunk 10]
```diff
@@ -43,8 +44,44 @@ impl<T: Trait> Default for ChangeEntry<T> {
...
 pub trait AccountDb<T: Trait> {
-	fn get_storage(&self, account: &T::AccountId, location: &[u8]) -> Option<Vec<u8>>;
+	fn get_subtrie(&self, account: &T::AccountId) -> Option<SubTrie>;
+	fn get_or_create_keyspace(&self, account: &T::AccountId) -> KeySpace;
+	// this function is borderline (impl specific to overlaydb). An alternate
+	// design is to use rc internally (instead of optional cache) for 
+	// overlaydb. It avoid querying the cache through a recursive call,
+	// but make `get_subtrie` and `get_or_create_keyspace` recursive call
+	// awkward : multiple query or insert (unless we specify top overlaydb instance).
+	fn get_account_from_reg_ks(&self, ks: &KeySpace) -> Option<T::AccountId>;
```
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/account_db.rs]: This is used for get_storage of overlaydb. Am I wrong if I say that `AccountDB` trait is not used out of this module? I believe it could be seen as a technical trait for the overlay+direct linked list of layers.
Switching to Rc could remove it.

File: srml/contract/src/account_db.rs [hunk 11]
```diff
@@ -43,8 +44,44 @@ impl<T: Trait> Default for ChangeEntry<T> {
 
 pub type ChangeSet<T> = BTreeMap<<T as system::Trait>::AccountId, ChangeEntry<T>>;
 
+#[derive(Clone, Default)]
+pub struct AccountKeySpaceMapping<A: Ord> {
```
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/account_db.rs]: It is a cache to avoid accessing subtrie data through direct db at every call to a subkey.
Storage got its own lru cache, so I am starting to wonder if it is totally usefull.
Could prove more usefull if it mapped account_id to `SubTrie` struct.

File: srml/contract/src/account_db.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (pepyakin) [reply srml/contract/src/account_db.rs]: > Am I wrong if I say that AccountDB trait is not used out of this module?

When I say "within this module" I mean a Rust module or ".rs" file. So yeah, the `AccountDb` trait is **used** outside of this module. However, this trait is not used outside of this crate.

File: srml/contract/src/account_db.rs (hunk 1 above)
Comment (cheme) [reply srml/contract/src/account_db.rs]: I firstly wrote the code with a single instance under a Rc RefCell, but I did not like the way caching would be done at every overlay layer, but it is only a matter of specializing the first layer for being the only one doing the caching (it would make the code more complex but I can still try switch back to that).

File: srml/contract/src/account_db.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/account_db.rs]: Oups, I indeed misunderstood, this method is bad design (as written in the code comment), but the way the linked account overlay list run I cannot really think of an alternative (except the Rc design).

File: srml/contract/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (pepyakin) [reply srml/contract/src/lib.rs]: I mean, why we can't delete it from here? It looks like this definition is not used anywhere and might be confusing.

I am so curious about this because we have a convention for using TODOs, basically we are trying to avoid them in Substrate and leave only those which are justified and have an issue in the issue tracker.

I'd say that this particular `StorageDoubleMap` instantiation is unused and thus useless (Note that `StorageDoubleMap` might be a useful primitive, and I'm not proposing to remove it).

File: srml/contract/src/tests.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (pepyakin) [reply srml/contract/src/tests.rs]: Hm, but `DummyContractAddressFor` and `DummyKeySpaceGenerator` have different requirements, no? As far as I understad, it is ok to implement `ContractAddressFor` and return a non-unique from `DummyContractAddressFor` simply because it doesn't require it. However, `KeySpaceGenerator` has this requirement which `DummyKeySpaceGenerator` doesn't satisfy.

It is not the end of the world, but personally I'd prefer keeping traits to actually satisfy the required contract. In the exceptional cases, if doing so is not possible/feasible, we need to at least leave a comment explaining that why we went this way. Otherwise it might be confusing for the future maintainers.

File: srml/contract/src/account_db.rs (hunk 1 above)
Comment (pepyakin) [reply srml/contract/src/account_db.rs]: > but I did not like the way caching would be done at every overlay layer

But now, it looks like it also done on each layer, no? The cache is modified for the each layer (i.e. grep `ka_mut.insert` or `ka.borrow_mut().insert`)?

Comment (pepyakin) [reply srml/contract/src/account_db.rs]: Ah, ok, I see now, everything is wrapped in a `map` and `keyspace_account` is `None` for the every layer except for the second (or in the other words, the first `OverlayAccountDb` layer).

Well, tbh, I think this is rather convoluted from the code perspective, and it sounds like a single cache instance should be a more simpler solution, no? why no?

File: srml/contract/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/lib.rs]: It is really not used, the TODO behing here to ask if we want to keep the functionnality, seems like we do not want to keep unused code, so let's remember f2bd95a9a53bd001d7fd988ecdfb486da88565e6 if we want to revert the deletion of this one.

[the remark this answers is not in the document]
Comment (pepyakin) [reply srml/contract/src/lib.rs]: Works for me!

File: srml/contract/src/tests.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/tests.rs]: The issue with DummyContractAddressFor is that we can only generate one accountid so it limits test cases.

[the remark this answers is not in the document]
Comment (pepyakin) [reply srml/contract/src/tests.rs]: Indeed, this is a downside of `DummyContractAddressFor`.

File: srml/contract/src/account_db.rs (hunk 1 above)
Comment (cheme) [reply srml/contract/src/account_db.rs]: It also need to have this kind of specialization otherwise insert of value from NativeAccount will be done at every OverlayAccount (rc will work super nice for access, not for insert). But it is only adding a bool at the first layer to ensure the update is only done under this condition, I can switch to it (there is an interest in term of perf).

File: srml/contract/src/account_db.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (pepyakin) [reply srml/contract/src/account_db.rs]: > Storage got its own lru cache

What do you mean? I can't recall anything about that...

[the remark this answers is not in the document]
Comment (cheme) [reply srml/contract/src/account_db.rs]: Some LruCache usage into core/client/db/src/storage_cache.rs , but with a `AccountInfo` struct we probably do not want to go down to this and deserialize each time.

**Comment by arkpar** [comment]
For avoiding DB conflicts we could probably use the same trick as parity-ethereum `AccoundDb` does. I.e. prefix or XOR each db key for contract trie with contract address.

File: srml/support/src/double_map.rs [hunk 12]
```diff
@@ -1,138 +0,0 @@
...
-///
-/// It provides an important ability to efficiently remove all entries
-/// that have a common first key.
-///
-/// # Mapping of keys to a storage path
-///
-/// The storage key (i.e. the key under which the `Value` will be stored) is created from two parts.
-/// The first part is a hash of a concatenation of the `PREFIX` and `Key1`. And the second part
-/// is a hash of a `Key2`.
-pub trait StorageDoubleMap {
```
Comment (pepyakin) [inline srml/support/src/double_map.rs]: As I've mentioned [in my comment](https://github.com/paritytech/substrate/pull/2002#discussion_r266493543), I don't see a reason for deleting `StorageDoubleMap`, it is just contracts don't use it anymore. In fact, there is [a PR](https://github.com/paritytech/substrate/pull/1918) that promotes `StorageDoubleMap` to be a first-class citizen in decl_storage. 

Or am I missing something?

Comment (cheme) [reply srml/support/src/double_map.rs]: My fault, I understood your comment as a request for removal, personally I wanted to keep it but considering it was not used I did the removal. (you probably wanted to remove the TODO, not the StorageDoubleMap :)
Seeing that it will be move by the cited PR, I can restore it (good thing I kept the commit hash) and also remove the TODO comment.

Comment (pepyakin) [reply srml/support/src/double_map.rs]: To be clear: I wanted to keep `StorageDoubleMap`, remove `StorageOf` (which implements `StorageDoubleMap`) because it is not needed, is it?

Comment (cheme) [reply srml/support/src/double_map.rs]: Not needed, should be restored at this point, I am not removing StorageOf in the PR as it will be redundant with #1918, thinking twice it will not conflict if I also remove it here: I will remove it in next commit.

File: srml/contract/src/lib.rs [hunk 13]
```diff
@@ -342,11 +386,14 @@ decl_storage! {
...
+		/// The subtrie counter
+		pub AccountCounter: u64 = 0;
+		/// The code associated with a given account.
+		pub AccountInfoOf: map T::AccountId => Option<AccountInfo>;
 	}
 }
 
-/// The storage items associated with an account/key.
-///
+/// The storage items associated with a prefix space in a tree
```
Comment (pepyakin) [inline srml/contract/src/lib.rs]: We need to remove `StorageOf`, as discussed [here](https://github.com/paritytech/substrate/pull/2002#discussion_r266944148) since it is not used.

File: srml/contract/src/lib.rs [hunk 14]
```diff
@@ -73,17 +73,18 @@ use rstd::marker::PhantomData;
...
-use srml_support::{Parameter, StorageMap, StorageValue, StorageDoubleMap, decl_module, decl_event, decl_storage};
+use srml_support::{Parameter, StorageMap, StorageValue, StorageDoubleMap, decl_module, decl_event, decl_storage, storage::child};
 use srml_support::traits::OnFreeBalanceZero;
 use system::{ensure_signed, RawOrigin};
 use runtime_io::{blake2_256, twox_128};
 use timestamp;
 use fees;
 
 pub type CodeHash<T> = <T as system::Trait>::Hash;
+pub type KeySpace = Vec<u8>;
```
Comment (pepyakin) [inline srml/contract/src/lib.rs]: I find the name `KeySpace` to be confusing, tbh. I remember you told me, that you chose that name because of some underlying details of the storage implementation. However, I think that the connection is somewhat convoluted.

I'd prefer leave the name `trie_id` as found in the original design doc. This would involve changing the other names as well.

File: srml/contract/src/tests.rs [hunk 15]
```diff
@@ -111,6 +117,17 @@ impl ContractAddressFor<H256, u64> for DummyContractAddressFor {
 	}
 }
 
+static KEY_COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+pub struct DummyKeySpaceGenerator;
+impl KeySpaceGenerator<u64> for DummyKeySpaceGenerator {
+	fn key_space(account_id: &u64) -> KeySpace {
+		let mut res = KEY_COUNTER.fetch_add(1,Ordering::Relaxed).to_le_bytes().to_vec();
```
Comment (pepyakin) [inline srml/contract/src/tests.rs]:
```suggestion
		let mut res = KEY_COUNTER.fetch_add(1, Ordering::Relaxed).to_le_bytes().to_vec();
```

File: srml/contract/src/lib.rs [hunk 16]
```diff
@@ -92,6 +93,47 @@ pub trait ComputeDispatchFee<Call, Balance> {
...
+pub struct KeySpaceFromParentCounter<T: Trait>(PhantomData<T>);
+
+/// This generator use inner counter for account id and apply hash over `AccountId +
+/// parent_keyspace + accountid_counter`
+impl<T: Trait> KeySpaceGenerator<T::AccountId> for KeySpaceFromParentCounter<T>
+where
+	T::AccountId: AsRef<[u8]> {
+	fn key_space(account_id: &T::AccountId) -> KeySpace {
+    // note that skipping a value due to error is not an issue here.
+    // we only need uniqueness, not sequence.
```
Comment (pepyakin) [inline srml/contract/src/lib.rs]: Can we change these to tabs?

```suggestion
	// we only need uniqueness, not sequence.
```

File: srml/contract/src/lib.rs [hunk 17]
```diff
@@ -92,6 +93,47 @@ pub trait ComputeDispatchFee<Call, Balance> {
...
+}
+
+/// Get key space from `account_id`
+pub struct KeySpaceFromParentCounter<T: Trait>(PhantomData<T>);
+
+/// This generator use inner counter for account id and apply hash over `AccountId +
+/// parent_keyspace + accountid_counter`
+impl<T: Trait> KeySpaceGenerator<T::AccountId> for KeySpaceFromParentCounter<T>
+where
+	T::AccountId: AsRef<[u8]> {
```
Comment (pepyakin) [inline srml/contract/src/lib.rs]:
```suggestion
	T::AccountId: AsRef<[u8]>
{
```

File: srml/contract/src/lib.rs [hunk 18]
```diff
@@ -92,6 +93,47 @@ pub trait ComputeDispatchFee<Call, Balance> {
...
+	fn key_space(account_id: &T::AccountId) -> KeySpace {
+    // note that skipping a value due to error is not an issue here.
```
Comment (pepyakin) [inline srml/contract/src/lib.rs]:
```suggestion
	// note that skipping a value due to error is not an issue here.
```

File: core/state-machine/src/lib.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (pepyakin) [reply core/state-machine/src/lib.rs]: So because it was moved in the issue and because it is unused and redundant, should we remove this function? What about the one below?

**Comment by pepyakin** [comment]
Rebase is needed. Also, @sorpaas / @arkpar can we have a lil bit more eyes to check the non-contract parts?

**Comment by cheme** [comment]
@pepyakin sorry for the lack of activity on this pr (was trying to get my head on the right way to isolate keyspace (prototyping on #1918)).
Currently how I see thing is that 'keyspace' will be additional data stored with the trie_root of the child trie (it needs to be at a lower level).
So changing the keyspace reference to trie_id in this pr makes sense.
Then initiating a child trie could be done with the trie_id as initial keyspace value (or seed).
At some point I wonder if we even need the trie_id (I will need a mechanism to generate keyspace at a lower level (unless I change the api a lot).

File: srml/contract/src/lib.rs (hunk 14 above)
Comment (cheme) [reply srml/contract/src/lib.rs]: As in my comment bellow, I ended up to similar conclusion (keyspace being inplementation details of child trie).

File: core/state-machine/src/ext.rs [hunk 19]
```diff
@@ -264,6 +264,21 @@ where
 		});
 	}
 
+	fn clear_child_prefix(&mut self, storage_key: &[u8], prefix: &[u8]) {
+		let _guard = panic_handler::AbortGuard::new(true);
+		if is_child_storage_key(prefix) {
```
Comment (sorpaas) [inline core/state-machine/src/ext.rs]: I don't quite understand this `if`. `prefix` is a key in child storage, so it has nothing to do with root's `is_child_storage_key`?

File: core/trie/src/lib.rs [hunk 20]
```diff
@@ -152,6 +162,21 @@ pub fn child_trie_root<H: Hasher, I, A, B>(_storage_key: &[u8], input: I) -> Vec
 	trie_root::<H, _, _, _>(input).as_ref().iter().cloned().collect()
 }
 
+pub fn child_trie_root2<H: Hasher, I, A, B>(_storage_key: &[u8], input: I) -> H::Out where
```
Comment (sorpaas) [inline core/trie/src/lib.rs]: Are this actually used? The reason for `child_trie_root` to return `Vec<u8>` is so that it can support multi-hash -- some hash algorithms applied to trie hash in the future might use a different length.

`H` is only the default hasher here, meaning a multi-hash if branch can ignore that generic parameter.

File: core/state-machine/src/trie_backend_essence.rs [hunk 21]
```diff
@@ -141,6 +141,48 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
...
+			}
+		};
+
+		let mut read_overlay = MemoryDB::default();
+		let eph = Ephemeral {
+			storage: &self.storage,
+			overlay: &mut read_overlay,
+		};
+
+		let mut iter = move || -> Result<(), Box<TrieError<H::Out>>> {
```
Comment (sorpaas) [inline core/state-machine/src/trie_backend_essence.rs]: I think this function has the same issue. We'd either:

* Remove support for variable length for multi-hash, that will make it so that we can use `H::Out` instead of `Vec<u8>` everywhere.
* Or take `root` as `Vec<u8>` here and move this iterator into `trie` struct.

File: srml/support/src/storage/mod.rs [hunk 22]
```diff
@@ -560,6 +576,109 @@ pub mod unhashed {
...
+	}
+
+	/// Ensure `key` has no explicit entry in storage.
+	pub fn kill(storage_key: &[u8], key: &[u8]) {
+		runtime_io::clear_child_storage(storage_key, key);
+	}
+
+	/// Ensure keys with the given `prefix` have no entries in storage.
+	pub fn kill_prefix(storage_key: &[u8], prefix: &[u8]) {
+		runtime_io::clear_child_prefix(storage_key, prefix);
```
Comment (sorpaas) [inline srml/support/src/storage/mod.rs]: I probably missed it, but do you mind to explain why we still need prefix for a child storage?

File: core/state-machine/src/ext.rs (hunk 19 above)
Comment (cheme) [reply core/state-machine/src/ext.rs]: I should remove this (all new child function) from the PR, it is not needed (I was making the implementation at somepoint because I wanted to generalize child trie logic to root trie to, but that is an open question (making root trie a child trie with parent trie &[]) -> this result to hierarchical trie addressing and is out of this pr scope.

File: core/trie/src/lib.rs (hunk 20 above)
Comment (cheme) [reply core/trie/src/lib.rs]: Not used anymore, I will remove it.

**Comment by pepyakin** [comment]
Hey @cheme, can you update this PR?

**Comment by cheme** [comment]
Sure (I need to make branch in parity repo when creating pr (i think you can edit my branch but you need to add the remote which is not super nice)).

**Comment by pepyakin** [comment]
Heh, I've tried actually : ) I used my ssh git access but I've got a permission denied error.

**Comment by cheme** [comment]
The ci did manage to do it at some point (the automatic wasm build), maybe it is not super stable :)
